File: sdk/go/templateroot_test.go
package shojiku

import (
	"context"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// generate runs the name entrance and insists the call itself succeeded — a
// refused NAME is a failed result, never an error, so a test that gets an
// error here has found a different bug from the one it was written for.
func generate(t *testing.T, client *Client, name string) *Result {
	t.Helper()
	result, err := client.Generate(context.Background(), name, map[string]any{})
	if err != nil {
		t.Fatalf("Generate(%q) returned an error rather than a failed result: %v", name, err)
	}
	return result
}

func assertRejected(t *testing.T, result *Result, kind string) {
	t.Helper()
	if result.Success() {
		t.Fatal("the name was accepted")
	}
	if got := result.Failure().Kind(); got != kind {
		t.Errorf("kind = %q, want %q (message: %s)", got, kind, result.Failure().Message())
	}
}

func TestABlankNameIsARefusedRequestRatherThanMisuse(t *testing.T) {
	// It can arrive straight from a form field, so it is a fact about the
	// request rather than a bug in the program.
	client := newTestClient(t)

	for _, name := range []string{"", "   ", "\t"} {
		assertRejected(t, generate(t, client, name), "template_name")
	}
}

func TestEveryHostileNameShapeIsRefused(t *testing.T) {
	// The union across platforms, not the host's: a name valid on one machine
	// is valid on all of them, which is the only way the same application
	// deploys to Linux and Windows both.
	client := newTestClient(t)
	for _, hostile := range []struct {
		name string
		why  string
	}{
		{"/etc/passwd", "an absolute path"},
		{"C:receipt", "drive-relative"},
		{`\\host\share`, "a UNC path"},
		{"../receipt", "traversal with a separator"},
		{"sub/receipt", "a forward-slash separator"},
		{`sub\receipt`, "a backslash separator"},
		{"receipt\x00.yml", "a NUL"},
		{"rec\x1feipt", "a control character"},
		{"CON", "a reserved device"},
		{"NUL", "a reserved device"},
		{"con.yml", "a reserved device with an extension"},
		{"CON.", "a reserved device with a trailing dot"},
		{"CON ", "a reserved device with a trailing space"},
	} {
		t.Run(hostile.why+"/"+hostile.name, func(t *testing.T) {
			assertRejected(t, generate(t, client, hostile.name), "template_name")
		})
	}
}

func TestBareTraversalIsRefusedByContainmentRatherThanByShape(t *testing.T) {
	// `..` carries no separator, so no name-shape rule fires on it. What
	// refuses it is canonicalization: the answer is outside the root.
	assertRejected(t, generate(t, newTestClient(t), ".."), "template_escapes_root")
}

func TestASymlinkPointingOutsideTheRootIsNotFollowed(t *testing.T) {
	outside := t.TempDir()
	if err := os.WriteFile(filepath.Join(outside, templateFile), []byte("version: 0.1.0\n"), 0o644); err != nil {
		t.Fatalf("writing the outside template: %v", err)
	}
	root := t.TempDir()
	if err := os.Symlink(outside, filepath.Join(root, "escape")); err != nil {
		t.Fatalf("linking: %v", err)
	}

	_, rejection := (&templateRoot{path: root}).resolve("escape")

	if rejection == nil || rejection.kind != "template_escapes_root" {
		t.Fatalf("rejection = %+v, want template_escapes_root", rejection)
	}
}

func TestASiblingSharingTheRootsPrefixDoesNotPassContainment(t *testing.T) {
	// The reason containment is a STRUCTURAL parent test rather than a string
	// prefix compare: `<root>-evil` starts with the root's own path.
	parent := t.TempDir()
	root := filepath.Join(parent, "root")
	evil := filepath.Join(parent, "root-evil")
	for _, dir := range []string{root, evil} {
		if err := os.MkdirAll(dir, 0o700); err != nil {
			t.Fatalf("creating %s: %v", dir, err)
		}
	}

	resolved, rejection := (&templateRoot{path: root}).contained(evil)

	if rejection == nil || rejection.kind != "template_escapes_root" {
		t.Fatalf("contained(%q) = %q, %+v; want it refused", evil, resolved, rejection)
	}
}

func TestARootThatDoesNotExistIsNotFound(t *testing.T) {
	root := &templateRoot{path: filepath.Join(t.TempDir(), "absent")}

	_, rejection := root.resolve("receipt")

	if rejection == nil || rejection.kind != "template_not_found" {
		t.Fatalf("rejection = %+v, want template_not_found", rejection)
	}
}

func TestANameThatDoesNotExistIsNotFound(t *testing.T) {
	assertRejected(t, generate(t, newTestClient(t), "no-such-template"), "template_not_found")
}

func TestADirectoryWithNoTemplateFileIsRefusedAsUnreadable(t *testing.T) {
	root := t.TempDir()
	if err := os.MkdirAll(filepath.Join(root, "empty"), 0o700); err != nil {
		t.Fatalf("creating the fixture: %v", err)
	}

	_, rejection := (&templateRoot{path: root}).resolve("empty")

	if rejection == nil || rejection.kind != "template_unreadable" {
		t.Fatalf("rejection = %+v, want template_unreadable", rejection)
	}
	if rejection.cause == "" {
		t.Error("the rejection carries no cause for the trace")
	}
}

func TestADirectoryWhereTheTemplateBelongsIsRefusedToo(t *testing.T) {
	// The structural form of "unreadable", which is the shape a gate running
	// as root can actually produce — a chmod proves nothing there.
	root := t.TempDir()
	if err := os.MkdirAll(filepath.Join(root, "odd", templateFile), 0o700); err != nil {
		t.Fatalf("creating the fixture: %v", err)
	}

	_, rejection := (&templateRoot{path: root}).resolve("odd")

	if rejection == nil || rejection.kind != "template_unreadable" {
		t.Fatalf("rejection = %+v, want template_unreadable", rejection)
	}
}

func TestAnOptionalDefinitionsFileIsUsedWhenPresentAndAbsentOtherwise(t *testing.T) {
	root := &templateRoot{path: fixtureTemplates(t)}

	withSchema, rejection := root.resolve("receipt")
	if rejection != nil {
		t.Fatalf("resolving receipt: %+v", rejection)
	}
	if !strings.HasSuffix(withSchema.definitions, definitionsFile) {
		t.Errorf("definitions = %q, want the fixture's own", withSchema.definitions)
	}

	without, rejection := root.resolve("warns")
	if rejection != nil {
		t.Fatalf("resolving warns: %+v", rejection)
	}
	if without.definitions != "" {
		t.Errorf("definitions = %q, want none", without.definitions)
	}
	if without.assetsDir == "" {
		t.Error("the assets directory is not set")
	}
}

func TestAHostileNameIsEchoedBoundedInItsRefusal(t *testing.T) {
	long := strings.Repeat("a", 200) + "/escape"

	result := generate(t, newTestClient(t), long)

	message := result.Failure().Message()
	if strings.Contains(message, strings.Repeat("a", textLimit+1)) {
		t.Errorf("the refusal echoed an unbounded name: %s", message)
	}
}

func TestGeneratingWithNoTemplateRootIsMisuse(t *testing.T) {
	// Not a failed result: nothing was asked about a document. The message
	// names every way to configure one.
	client, err := NewClient(WithEnv(false), WithBinary(engineBinary(t)))
	if err != nil {
		t.Fatalf("building the client: %v", err)
	}

	_, err = client.Generate(context.Background(), "receipt", nil)

	assertUsage(t, err, "WithTemplates")
	if client.TemplateRoot() != "" {
		t.Errorf("TemplateRoot() = %q, want empty", client.TemplateRoot())
	}
}

File: sdk/php/tests/TemplateRootTest.php
<?php

declare(strict_types=1);

namespace Shojiku\Tests;

use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\TestCase;
use Shojiku\Client;
use Shojiku\Configuration;
use Shojiku\Exception\UsageException;
use Shojiku\Step;
use Shojiku\TemplateRoot;

/**
 * The hardening, one test per claim.
 *
 * The rejection rules are the UNION across platforms, so every one of these
 * is refused on this machine too — a template name that is valid on one
 * machine has to be valid on all of them, which is the only way the same
 * application deploys to both.
 *
 * Hostile control characters are written as PHP ESCAPES, never as raw bytes:
 * a raw one makes the whole file binary-classified and silently absent from
 * every `grep`.
 */
final class TemplateRootTest extends TestCase
{
    use EngineFixtures;

    /**
     * @return array<string, array{string}>
     */
    public static function refusedNames(): array
    {
        return [
            'empty' => [''],
            'blank' => ['   '],
            'absolute posix' => ['/etc/passwd'],
            'traversal' => ['../receipt'],
            'nested' => ['business/receipt'],
            'backslash separator' => ['business\\receipt'],
            'windows absolute' => ['\\\\host\\share'],
            'drive relative' => ['C:receipt'],
            'nul byte' => ["recei\x00pt"],
            'escape byte' => ["recei\x1bpt"],
            'del byte' => ["recei\x7fpt"],
            'reserved device' => ['CON'],
            'reserved device lowercase' => ['nul'],
            'reserved device with extension' => ['CON.yml'],
            'reserved device with trailing dot' => ['CON.'],
            'reserved device with trailing space' => ['CON '],
            'reserved port device' => ['COM1'],
        ];
    }

    #[DataProvider('refusedNames')]
    public function testAHostileNameIsARefusedRequestRatherThanAnException(string $name): void
    {
        $result = $this->client()->generate($name, []);

        self::assertTrue($result->failed(), sprintf('`%s` was not refused', addcslashes($name, "\0..\37")));
        $failure = $result->failure();
        self::assertNotNull($failure);
        self::assertSame('template_name', $failure->kind());
        self::assertSame(Step::Generate, $failure->step());
    }

    public function testAMissingNameIsRefusedAsNotFound(): void
    {
        $result = $this->client()->generate('no-such-template', []);

        $failure = $result->failure();
        self::assertNotNull($failure);
        self::assertSame('template_not_found', $failure->kind());
    }

    public function testADirectoryWithoutATemplateFileIsRefusedAsUnreadable(): void
    {
        $this->inTempDir(function (string $root): void {
            mkdir($root.'/empty');
            // A directory where `templates.yml` belongs, not a chmod: the gate
            // container runs as ROOT, and root ignores permission bits, so a
            // `chmod 000` fixture would pass for the wrong reason.
            mkdir($root.'/empty/templates.yml');

            $failure = $this->client(['templates' => $root])->generate('empty', [])->failure();

            self::assertNotNull($failure);
            self::assertSame('template_unreadable', $failure->kind());
            self::assertNotNull($failure->cause());
            self::assertSame('io', $failure->cause()?->kind());
        });
    }

    public function testASymlinkOutOfTheRootIsRefusedAfterCanonicalization(): void
    {
        $this->inTempDir(function (string $dir): void {
            mkdir($dir.'/root');
            mkdir($dir.'/outside');
            file_put_contents($dir.'/outside/templates.yml', "version: 0.1.0\nname: x\n");
            symlink($dir.'/outside', $dir.'/root/escape');

            $failure = $this->client(['templates' => $dir.'/root'])->generate('escape', [])->failure();

            self::assertNotNull($failure);
            self::assertSame('template_escapes_root', $failure->kind());
        });
    }

    public function testASiblingDirectorySharingTheRootsPrefixIsNotInsideIt(): void
    {
        // A prefix compare would accept `<root>-evil`; the containment test is
        // structural for exactly this case.
        $this->inTempDir(function (string $dir): void {
            mkdir($dir.'/root');
            mkdir($dir.'/root-evil');
            file_put_contents($dir.'/root-evil/templates.yml', "version: 0.1.0\nname: x\n");
            symlink($dir.'/root-evil', $dir.'/root/evil');

            $failure = $this->client(['templates' => $dir.'/root'])->generate('evil', [])->failure();

            self::assertNotNull($failure);
            self::assertSame('template_escapes_root', $failure->kind());
        });
    }

    public function testAMissingRootIsRefusedRatherThanCompared(): void
    {
        // `realpath()` returns FALSE for a path that does not exist rather
        // than raising; a containment check that skipped that branch would
        // compare a false against a string.
        $failure = $this->client(['templates' => '/nonexistent-root'])->generate('receipt', [])->failure();

        self::assertNotNull($failure);
        self::assertSame('template_not_found', $failure->kind());
    }

    public function testANameThatIsNotAStringIsProgrammerMisuse(): void
    {
        $this->expectException(UsageException::class);
        $this->expectExceptionMessageMatches('/must be a string; got int/');
        $this->client()->generate(42, []);
    }

    public function testTheRejectionMessageBoundsWhatItEchoes(): void
    {
        $failure = $this->client()->generate(str_repeat('a', 500).'/x', [])->failure();

        self::assertNotNull($failure);
        // Bounded to the echo limit plus the message around it — what matters
        // is that a hostile name cannot make the message unbounded.
        self::assertLessThan(300, strlen($failure->message()));
    }

    public function testTheRootIsResolvedThroughTheConfiguredValueFirst(): void
    {
        $root = new TemplateRoot($this->fixtureTemplates());

        self::assertSame($this->fixtureTemplates(), $root->path());
        $sources = $root->resolve('receipt');
        self::assertStringEndsWith('receipt/templates.yml', $sources->template);
        self::assertStringEndsWith('receipt/definitions.yml', (string) $sources->definitions);
        self::assertStringEndsWith('receipt', (string) $sources->assetsDir);
    }

    public function testAnOptionalDefinitionsFileIsAbsentRatherThanEmpty(): void
    {
        $sources = (new TemplateRoot($this->fixtureTemplates()))->resolve('warns');

        self::assertNull($sources->definitions);
    }

    public function testAnExplicitRootBeatsConfigurationWhichBeatsTheEnvironment(): void
    {
        putenv('SHOJIKU_TEMPLATE_ROOT=/from-environment');

        try {
            // The environment supplies it when nothing else does…
            self::assertSame('/from-environment', (new Client())->templateRoot()?->path());
            // …configuration beats the environment…
            Configuration::configure(['templates' => '/configured']);
            self::assertSame('/configured', (new Client())->templateRoot()?->path());
            // …an explicit argument beats both…
            self::assertSame('/explicit', (new Client(templates: '/explicit'))->templateRoot()?->path());
            // …and one flag turns every `SHOJIKU_*` lookup off at once. The
            // binary is passed explicitly for the same reason: with lookups
            // off, `SHOJIKU_BIN` is not read either, and a client cannot be
            // built over an engine it cannot find.
            Configuration::reset();
            self::assertNull((new Client(binary: $this->engineBinary(), env: false))->templateRoot());
        } finally {
            putenv('SHOJIKU_TEMPLATE_ROOT');
        }
    }

    // The SHAPE of the root itself. Everything above constrains the NAME; what a
    // root may look like was never pinned, and the .NET SDK drifted there — its
    // canonical form kept a trailing separator while the parents it compared
    // against did not, so `templates/` could never contain anything. PHP is
    // immune because realpath() drops the separator; these pin that rather than
    // leaving it to the implementation it happens to use.

    /**
     * @return iterable<string, array{string}>
     */
    public static function rootSuffixes(): iterable
    {
        yield 'no separator' => [''];
        yield 'one separator' => ['/'];
        yield 'repeated separators' => ['//'];
    }

    #[DataProvider('rootSuffixes')]
    public function testATrailingSeparatorOnTheRootStillResolves(string $suffix): void
    {
        $result = $this->client(['templates' => $this->fixtureTemplates().$suffix])->generate('receipt', []);

        self::assertTrue($result->success(), (string) $result->failure()?->message());
    }

    public function testARelativeRootResolves(): void
    {
        // chdir is process-global, so it is restored in a finally rather than
        // left for the next test to discover.
        $previous = getcwd();
        self::assertIsString($previous);
        try {
            chdir(dirname($this->fixtureTemplates()));
            $name = basename($this->fixtureTemplates());

            foreach ([$name, $name.'/'] as $relative) {
                $result = $this->client(['templates' => $relative])->generate('receipt', []);

                self::assertTrue($result->success(), $relative.': '.((string) $result->failure()?->message()));
            }
        } finally {
            chdir($previous);
        }
    }

    public function testATrailingSeparatorDoesNotFollowASymlinkOutOfTheRoot(): void
    {
        // Normalizing the root must not loosen containment.
        $this->inTempDir(function (string $dir): void {
            mkdir($dir.'/root');
            mkdir($dir.'/outside');
            file_put_contents($dir.'/outside/templates.yml', "version: 0.1.0\nname: x\n");
            symlink($dir.'/outside', $dir.'/root/escape');

            $failure = $this->client(['templates' => $dir.'/root/'])->generate('escape', [])->failure();

            self::assertNotNull($failure);
            self::assertSame('template_escapes_root', $failure->kind());
        });
    }

    public function testATrailingSeparatorDoesNotAdmitAPrefixSibling(): void
    {
        // A string prefix compare would accept `<root>-evil`; containment is
        // structural. Normalizing the root is what makes that mistake reachable.
        $this->inTempDir(function (string $dir): void {
            mkdir($dir.'/root-evil/receipt', 0o700, true);
            file_put_contents($dir.'/root-evil/receipt/templates.yml', "version: 0.1.0\nname: x\n");
            mkdir($dir.'/root');
            symlink($dir.'/root-evil/receipt', $dir.'/root/receipt');

            $failure = $this->client(['templates' => $dir.'/root/'])->generate('receipt', [])->failure();

            self::assertNotNull($failure);
            self::assertSame('template_escapes_root', $failure->kind());
        });
    }
}
